File: backend/user-service/src/main/java/com/grab/user/service/UserService.java
package com.grab.user.service;

import com.grab.user.dto.UpdateProfileRequest;
import com.grab.user.dto.UserProfileResponse;
import com.grab.user.exception.ResourceNotFoundException;
import com.grab.user.model.User;
import com.grab.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;

    public UserProfileResponse getProfile(UUID userId) {
        User user = findUserById(userId);
        return toResponse(user);
    }

    @Transactional
    public UserProfileResponse updateProfile(UUID userId, UpdateProfileRequest request) {
        User user = findUserById(userId);

        if (request.getName() != null) {
            user.setName(request.getName());
        }
        if (request.getEmail() != null) {
            if (userRepository.existsByEmail(request.getEmail())) {
                throw new IllegalArgumentException("Email already in use");
            }
            user.setEmail(request.getEmail());
        }
        if (request.getProfileImage() != null) {
            user.setProfileImage(request.getProfileImage());
        }
        if (request.getFcmToken() != null) {
            user.setFcmToken(request.getFcmToken());
        }

        return toResponse(userRepository.save(user));
    }

    @Transactional
    public void deactivateAccount(UUID userId) {
        User user = findUserById(userId);
        user.setIsActive(false);
        userRepository.save(user);
        log.info("User {} deactivated", userId);
    }

    private User findUserById(UUID userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found: " + userId));
    }

    private UserProfileResponse toResponse(User user) {
        return UserProfileResponse.builder()
                .id(user.getId())
                .name(user.getName())
                .phone(user.getPhone())
                .email(user.getEmail())
                .role(user.getRole().name())
                .profileImage(user.getProfileImage())
                .isActive(user.getIsActive())
                .createdAt(user.getCreatedAt())
                .build();
    }
}

File: backend/user-service/internal/service/user_service.go
package service

import (
	"errors"

	"github.com/deliverx/user-service/internal/model"
	"github.com/deliverx/user-service/internal/repository"
	"github.com/google/uuid"
)

type UpdateProfileRequest struct {
	Name         string  `json:"name"`
	Email        *string `json:"email"`
	ProfileImage *string `json:"profile_image"`
	FcmToken     *string `json:"fcm_token"`
}

type UserService struct {
	repo *repository.UserRepository
}

func NewUserService(repo *repository.UserRepository) *UserService {
	return &UserService{repo: repo}
}

func (s *UserService) GetProfile(userID uuid.UUID) (*model.User, error) {
	return s.repo.FindByID(userID)
}

func (s *UserService) UpdateProfile(userID uuid.UUID, req UpdateProfileRequest) (*model.User, error) {
	user, err := s.repo.FindByID(userID)
	if err != nil {
		return nil, errors.New("user not found")
	}

	if req.Name != "" {
		user.Name = req.Name
	}
	if req.Email != nil {
		user.Email = req.Email
	}
	if req.ProfileImage != nil {
		user.ProfileImage = req.ProfileImage
	}
	if req.FcmToken != nil {
		user.FcmToken = req.FcmToken
	}

	if err := s.repo.Update(user); err != nil {
		return nil, err
	}
	return user, nil
}

func (s *UserService) GetDrivers() ([]model.User, error) {
	return s.repo.FindDrivers()
}
